camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/DslDeferredFunctionCall.java

```diff
@@ -99,48 +99,81 @@ public Object call() throws Exception {
     }
 
     protected static Maybe<Object> invokeOn(Object obj, String fnName, List<?> args) {
-        Object instance = obj;
-        List<?> instanceArgs = args;
-        Maybe<Method> method = Reflections.getMethodFromArgs(instance, fnName, instanceArgs);
-
-        if (method.isAbsent()) {
+        return new Invoker(obj, fnName, args).invoke();
+    }
+    
+    protected static class Invoker {
+        final Object obj;
+        final String fnName;
+        final List<?> args;
+        
+        Maybe<Method> method;
+        Object instance;
+        List<?> instanceArgs;
+        
+        protected Invoker(Object obj, String fnName, List<?> args) {
+            this.fnName = fnName;
+            this.obj = obj;
+            this.args = args;
+        }
+        
+        protected Maybe<Object> invoke() {
+            findMethod();
+            
+            if (method.isPresent()) {
+                Method m = method.get();
+    
+                checkCallAllowed(m);
+    
+                try {
+                    // Value is most likely another BrooklynDslDeferredSupplier - let the caller handle it,
+                    return Maybe.of(Reflections.invokeMethodFromArgs(instance, m, instanceArgs));
+                } catch (IllegalArgumentException | IllegalAccessException | InvocationTargetException e) {
+                    // If the method is there but not executable for whatever reason fail with a fatal error, don't return an absent.
+                    throw Exceptions.propagate(new InvocationTargetException(e, "Error invoking '"+fnName+instanceArgs+"' on '"+instance+"'"));
```

Comment:
Minor: I don't think this should be an `InvocationTargetException` - `e` was not "thrown by an invoked method or constructor". If `e` is an IllegalArgumentException or IllegalAccessException, then it was caused by trying to invoke it; if `e` is an `InvocationTargetException` then it is already wrapped. We probably just want `Exceptions.propagateAnnotated("...", e)`